sql/src/main/java/org/apache/druid/sql/calcite/expression/Expressions.java

```diff
@@ -306,61 +309,79 @@ private static DruidExpression rexCallToDruidExpression(
-      return DruidExpression.ofLiteral(
-          columnType,
-          number == null ? DruidExpression.nullLiteral() : DruidExpression.longLiteral(number.longValue())
-      );
+      retVal = ExprEval.ofType(ExpressionType.LONG, number == null ? null : number.longValue());
     } else if (SqlTypeName.NUMERIC_TYPES.contains(sqlTypeName)) {
       // Numeric, non-INT, means we represent it as a double.
       final Number number = (Number) RexLiteral.value(rexNode);
-      return DruidExpression.ofLiteral(
-          columnType,
-          number == null ? DruidExpression.nullLiteral() : DruidExpression.doubleLiteral(number.doubleValue())
-      );
+      retVal = ExprEval.ofType(ExpressionType.DOUBLE, number == null ? null : number.doubleValue());
     } else if (SqlTypeFamily.INTERVAL_DAY_TIME == sqlTypeName.getFamily()) {
       // Calcite represents DAY-TIME intervals in milliseconds.
       final long milliseconds = ((Number) RexLiteral.value(rexNode)).longValue();
-      return DruidExpression.ofLiteral(columnType, DruidExpression.longLiteral(milliseconds));
+      retVal = ExprEval.ofType(ExpressionType.LONG, milliseconds);
     } else if (SqlTypeFamily.INTERVAL_YEAR_MONTH == sqlTypeName.getFamily()) {
       // Calcite represents YEAR-MONTH intervals in months.
       final long months = ((Number) RexLiteral.value(rexNode)).longValue();
-      return DruidExpression.ofLiteral(columnType, DruidExpression.longLiteral(months));
+      retVal = ExprEval.ofType(ExpressionType.LONG, months);
     } else if (SqlTypeName.STRING_TYPES.contains(sqlTypeName)) {
-      return DruidExpression.ofStringLiteral(RexLiteral.stringValue(rexNode));
+      final String s = RexLiteral.stringValue(rexNode);
+      retVal = ExprEval.ofType(ExpressionType.STRING, s);
+      isEmptyString = s != null && s.isEmpty();
     } else if (SqlTypeName.TIMESTAMP == sqlTypeName || SqlTypeName.DATE == sqlTypeName) {
-      if (RexLiteral.isNullLiteral(rexNode)) {
-        return DruidExpression.ofLiteral(columnType, DruidExpression.nullLiteral());
-      } else {
-        return DruidExpression.ofLiteral(
-            columnType,
-            DruidExpression.longLiteral(
-                Calcites.calciteDateTimeLiteralToJoda(rexNode, plannerContext.getTimeZone()).getMillis()
-            )
-        );
-      }
-    } else if (SqlTypeName.BOOLEAN == sqlTypeName) {
-      return DruidExpression.ofLiteral(
-          columnType,
-          DruidExpression.longLiteral(RexLiteral.booleanValue(rexNode) ? 1 : 0)
+      retVal = ExprEval.ofType(
+          ExpressionType.LONG,
+          Calcites.calciteDateTimeLiteralToJoda(rexNode, plannerContext.getTimeZone()).getMillis()
       );
+    } else if (SqlTypeName.BOOLEAN == sqlTypeName) {
+      retVal = ExprEval.ofType(ExpressionType.LONG, RexLiteral.booleanValue(rexNode) ? 1 : 0);
     } else {
       // Can't translate other literals.
       return null;
     }
+
+    return new ExprEvalWrapper(retVal, isEmptyString);
```

Comment:
The problem I'm trying to solve with `ExprEvalWrapper` is that in replace-with-default mode, there is no way to have a constant `Expr` or `ExprEval` that contains an empty string. Both `StringExpr` and `StringExprEval` convert empty strings to nulls in their constructors. So the type `ExprEval` cannot be used here, it has to be some kind of wrapper.

Another possible approach is to extend `ExprEval` with a special `EmptyStringExprEval` that is only used by the SQL planner, and return that whenever we're in default-value mode and the `trueValue` (or `actualValue`) is an empty string. But I didn't do it that way, because that seemed even worse somehow.

Reply:
I feel like we are making an exception to literals which are `null` - to work like they are true `null`-s in `defaultMode` ; however if you insert a `null` into a table - it will act like the empty string....

wouldn't it be possible to handle `null` values in case we are in `defaults-mode` as values; by making logic-s like [this](https://github.com/apache/druid/pull/16311/files#diff-172d30aafbdf15cef24fbd77b3293d7c215c30a3de6253f5139369ac0e15780aR3766) will become restricted with `Nullhandling.sqlCompatible()`?

Reply:
I'm not sure exactly what you mean, but I thought about the wrapping a bit more and realized there was no real need to use `ExprEval` at all. In the latest patch I replaced the `ExprEvalWrapper` with a `DruidLiteral` that is something like an `ExprEval`, but is able to differentiate `null` and `''`. Please let me know what you think.